### DISCUSSION TITLE
Generic polymorphism

### DISCUSSION THREAD

**Discussion by RomanSoloweow** [dotnet/csharplang#5408] [Language Ideas]
What prevents from implementing polymorphism based on types in generics?
The passed type is known even before compilation.

```C#
public static TResource? GetClassResource<TResource>(this IResourceNode resourceNode, string key)
where TResource:class
{
    TResource? value = null;
    return value;
}

public static TResource? GetStructResource<TResource>(this IResourceNode resourceNode, string key)
    where TResource:struct
{
    TResource? value = null;
    return value;
}
```

In the current implementation, it is not possible to implement a method that returns null for both classes and structures.

**Comment by svick** [comment]
> What prevents from implementing polymorphism based on types in generics?

Quite often, the main answer to that question is: the maintainers of the language have to decide this is the feature they should be working on next. Which features are added to the language is primarily about priorities, not technical hurdles.

> In the current implementation, it is not possible to implement a method that returns null for both classes and structures.

It's far from ideal and it may not be what you're looking for, but it is kind of possible:

```c#
public static TResource? GetStructResource<TResource>(this IResourceNode resourceNode, string key)
{
    if (default(TResource) is not null)
        throw new InvalidOperationException("This method does not support non-nullable value types");

    return default;
}

resourceNode.GetStructResource<string>("") // null
resourceNode.GetStructResource<int?>("") // null
resourceNode.GetStructResource<int>("") // exception
```

**Reply by jnm2** [reply]
@svick Can't `typeof(TResource).IsValueType && Nullable.GetUnderlyingType(typeof(TResource)) is null` be replaced by `default(TResource) is not null`?

**Reply by svick** [reply]
@jnm2 You're right, it can, I did not think of that.

**Comment by koszeggy** [comment]
The problem is that `TResource?` means different types with and without the `where struct` constraint:

```cs
// GetStructResource<int> has a return type of Nullable<int>
public static TResource? GetStructResource<TResource>(this IResourceNode resourceNode, string key)
    where TResource : struct
{
    return null;
}

// GetAnyResource<int> can only return int, you must use GetAnyResource<int?> to be able to return null
public static TResource? GetAnyResource<TResource>(this IResourceNode resourceNode, string key)
{
    return default;
}
```

See also my [SO answer](https://stackoverflow.com/a/69212362/5114784) to a somewhat related question for a more detailed explanation.
